asset-transfer-private-data/application-gateway-go/app.go

```diff
@@ -0,0 +1,358 @@
+
+	// Collection names.
+	org1PrivateCollectionName = "Org1MSPPrivateCollection"
+	org2PrivateCollectionName = "Org2MSPPrivateCollection"
+
+	Red   = "\033[31m"
+	Reset = "\033[0m"
+)
+
+// Use a unique key so that we can run multiple times.
+var now = time.Now()
+var assetID1 = fmt.Sprintf("asset%d", now.Unix())
+var assetID2 = fmt.Sprintf("asset%d", now.Unix()+1)
+
+func main() {
+	clientOrg1 := newGrpcConnection(
+		tlsCertPathOrg1,
+		peerEndpointOrg1,
+		peerNameOrg1,
+	)
+	defer clientOrg1.Close()
+
+	gatewayOrg1, err := client.Connect(
+		newIdentity(certDirectoryPathOrg1, mspIDOrg1),
+		client.WithSign(newSign(keyDirectoryPathOrg1)),
+		client.WithClientConnection(clientOrg1),
+		client.WithHash(hash.SHA256),
+	)
+	if err != nil {
+		panic(err)
+	}
+	defer gatewayOrg1.Close()
+
+	clientOrg2 := newGrpcConnection(
+		tlsCertPathOrg2,
+		peerEndpointOrg2,
+		peerNameOrg2,
+	)
+	defer clientOrg2.Close()
+
+	gatewayOrg2, err := client.Connect(
+		newIdentity(certDirectoryPathOrg2, mspIDOrg2),
+		client.WithSign(newSign(keyDirectoryPathOrg2)),
+		client.WithClientConnection(clientOrg2),
+		client.WithHash(hash.SHA256),
+	)
+	if err != nil {
+		panic(err)
+	}
+	defer gatewayOrg2.Close()
+
+	// Get the smart contract as an Org1 client.
+	contractOrg1 := gatewayOrg1.GetNetwork(channelName).GetContract(chaincodeName)
+
+	// Get the smart contract as an Org1 client.
+	contractOrg2 := gatewayOrg2.GetNetwork(channelName).GetContract(chaincodeName)
+
+	fmt.Println("~~~~~~~~~~~~~~~~ As Org1 Client ~~~~~~~~~~~~~~~~")
+
+	// Create new assets on the ledger.
+	createAssets(*contractOrg1)
+
+	// Read asset from the Org1's private data collection with ID in the given range.
+	getAssetByRange(*contractOrg1)
+
+	// Attempt to transfer asset without prior approval from Org2, transaction expected to fail.
+	fmt.Println("\nAttempt TransferAsset without prior AgreeToTransfer")
+	err = transferAsset(*contractOrg1, assetID1)
+	if err == nil {
+		doFail("TransferAsset transaction succeeded when it was expected to fail")
+	}
+	fmt.Printf("*** Received expected error: %+v\n", err)
+
+	fmt.Println("\n~~~~~~~~~~~~~~~~ As Org2 Client ~~~~~~~~~~~~~~~~")
+
+	// Read the asset by ID.
+	readAssetByID(*contractOrg2, assetID1)
+
+	// Make agreement to transfer the asset from Org1 to Org2.
+	agreeToTransfer(*contractOrg2, assetID1)
+
+	fmt.Println("\n~~~~~~~~~~~~~~~~ As Org1 Client ~~~~~~~~~~~~~~~~")
+
+	// Read transfer agreement.
+	readTransferAgreement(*contractOrg1, assetID1)
+
+	// Transfer asset to Org2.
+	transferAsset(*contractOrg1, assetID1)
+
+	// Again ReadAsset: results will show that the buyer identity now owns the asset.
+	readAssetByID(*contractOrg1, assetID1)
+
+	// Confirm that transfer removed the private details from the Org1 collection.
+	if org1ReadSuccess := readAssetPrivateDetails(*contractOrg1, assetID1, org1PrivateCollectionName); !org1ReadSuccess {
+		doFail(fmt.Sprintf("Asset private data still exists in %s", org1PrivateCollectionName))
+	}
```

Comment:
The logic is inverted here. We expect this call to fail since the private data is no longer present in the Org1 collection. 

I'm not sure whether it helps readability to have the additional (boolean) variable here. Your choice.

```suggestion
	if readAssetPrivateDetails(contractOrg1, assetID1, org1PrivateCollectionName) {
		doFail(fmt.Sprintf("Asset private data still exists in %s", org1PrivateCollectionName))
	}
```

Reply:
Yes, you're right. Fixed the mistake. I left the variable in there as it is also used in the typescript sample [app.ts:108](https://github.com/twoGiants/fabric-samples/blob/b5c18e2c93037736bf5030ec831aa719031f04f4/asset-transfer-private-data/application-gateway-typescript/src/app.ts#L108-L109). I changed the if statement to be like the typescript one. I think: `if org1ReadSuccess` reads better and is clearer than `if readAssetPrivateDetails`.